### PR TITLE
Migration: Jenkins -> Semaphore

### PR DESCRIPTION
**Pull request by Muen342**
Migrated jenkins to semaphore
- Added Build, Compile, Validations, Publish, Upgrade step
- compile-validate runs 
```
./retry_zinc ./gradlew clean publishToMavenLocal build -x test --no-daemon --stacktrace -PxmlSpotBugsReport=true
```
- check-scala-compatibility runs 
```
./retry_zinc ./gradlew clean build -x test --no-daemon --stacktrace -PxmlSpotBugsReport=true -PscalaVersion=2.12
```
- publishStep was migrated to 
```
 ./gradlewAll -PmavenUrl=$mavenUrl -PkeepAliveMode=session uploadArchives` after setting mavenUrl to the vault secret of ` v1/ci/kv/gradle/artifactory_snapshots_settings` or `v1/ci/kv/gradle/artifactory_preview_release_settings
```
- migrated downstream build to 
```
sem-trigger -p $project -b $DOWNSTREAM_BRANCH_NAME -f .semaphore/semaphore.yml
``` 
where DOWNSTREAM_BRANCH_NAME is set with `set_downstream_branch.sh`
- migrated the unit and integration tests to 
```
'./gradlew 
                unitTest integrationTest --no-daemon --stacktrace --continue 
                -PtestLoggingEvents=started,passed,skipped,failed -PmaxParallelForks=4 
                -PignoreFailures=true -PmaxTestRetries=1 -PmaxTestRetryFailures=5'
```
- uploaded results with 
```
shopt -s globstar && test-results publish --name Test-Suite --trim-output-to 1024 --omit-output-for-passed **/build/test-results/**/TEST-*.xml
```

Service bot already ran with codeartifact in service.yml
https://semaphore.ci.confluent.io/jobs/b6fa3e0e-f2e3-4fae-b569-add5a54078a3#L112

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: .semaphore/semaphore.yml [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  cancel:
+    when: "true"
+execution_time_limit:
+  hours: 4
+queue:
+  - when: "branch != 'master' and branch !~ '[0-9]+\\.[0-9]+\\.x'"
+    processing: parallel
+global_job_config:
+  env_vars:
+    - name: NO_OF_TEST_WORKER_NODES
```
Comment (santhoshct) [inline .semaphore/semaphore.yml:31]: why do you all of these env vars? It seems all are copied from ce-kafka. you can only use whichever you want and remove the rest

File: .semaphore/semaphore.yml [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        commands:
+          - make init-ci
+      jobs:
+        - name: Build, Compile, Validations, Publish, Upgrade
+          commands:
+            - >-
+              if [ "$NANO_VERSION" = "true" ] && [ "$RELEASE_JOB" = "false" ] && [ "$ENABLE_PUBLISH_ARTIFACTS" = "true" ]; then
+                . ci-tools ci-update-version
+              fi
+            - make compile-validate
```
Comment (santhoshct) [inline .semaphore/semaphore.yml:104]: do you need these checks. double check again with existing jenkins code.

File: .semaphore/semaphore.yml [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            - make check-scala-compatibility
+            - |
+              if [ "$PUBLISH" = "true" ] && [ "$SEMAPHORE_GIT_REF_TYPE" != "pull-request" ] && [ "$ENABLE_PUBLISH_ARTIFACTS" = "true" ]; then \
```
Comment (santhoshct) [inline .semaphore/semaphore.yml:107]: ensure these conditions are proper and relevant.

File: .semaphore/semaphore.yml [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                elif [[ "$SEMAPHORE_GIT_BRANCH" == *-alpha* ]]; then \
+                  mavenUrl=$(vault kv get v1/ci/kv/gradle/artifactory_preview_release_settings | grep mavenUrl | cut -d "," -f 2 | cut -d "'" -f 2); \
+                fi; \
+                ./gradlewAll -PmavenUrl=$mavenUrl -PkeepAliveMode=session uploadArchives; \
+              fi
+            - |
+              echo "PUBLISH: $PUBLISH, RELEASE_JOB: $RELEASE_JOB, SEMAPHORE_GIT_REF_TYPE: $SEMAPHORE_GIT_REF_TYPE, ENABLE_DOWNSTREAM_TRIGGER: $ENABLE_DOWNSTREAM_TRIGGER"
+              if [ "$PUBLISH" = "true" ] && [ "$RELEASE_JOB" = "false" ] && [ "$SEMAPHORE_GIT_REF_TYPE" != "pull-request" ] && [ "$ENABLE_DOWNSTREAM_TRIGGER" = "true" ]; then
+                for project in $DOWNSTREAM_PROJECTS; do
+                  sem-trigger -p $project -b $DOWNSTREAM_BRANCH_NAME -f .semaphore/semaphore.yml
```
Comment (santhoshct) [inline .semaphore/semaphore.yml:120]: is this required for open source kafka?

File: .semaphore/semaphore.yml (hunk 1 above)
Comment (Muen342) [reply .semaphore/semaphore.yml:31]: I'll remove the ones that aren't required

File: .semaphore/semaphore.yml (hunk 2 above)
Comment (Muen342) [reply .semaphore/semaphore.yml:104]: same for check scala
https://github.com/confluentinc/kafka/pull/1204/files#diff-e6ffa5dc854b843b3ee3c3c28f8eae2f436c2df2b1ca299cca1fa5982e390cf8L72-L73

File: .semaphore/semaphore.yml (hunk 3 above)
Comment (Muen342) [reply .semaphore/semaphore.yml:107]: yes, it maps directly to 
```
if (config.publish) {
      stage("Publish to artifactory") {
        if (!config.isReleaseJob && !config.isPrJob) {
            ciTool("ci-push-tag ${env.WORKSPACE} kafka")
        }

        if (config.isDevJob) {
          publishStep('artifactory_snapshots_settings')
        } else if (config.isPreviewJob) {
          publishStep('artifactory_preview_release_settings')
        }
      }
    }
```
from jenkinsfile

File: .semaphore/semaphore.yml (hunk 2 above)
Comment (Muen342) [reply .semaphore/semaphore.yml:94]: for CI update version, it's migrated from this 
https://github.com/confluentinc/kafka/pull/1204/files#diff-e6ffa5dc854b843b3ee3c3c28f8eae2f436c2df2b1ca299cca1fa5982e390cf8L66-L68

File: .semaphore/semaphore.yml [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        always:
+          commands:
+            - shopt -s globstar && test-results publish --name Test-Suite --trim-output-to 1024 --omit-output-for-passed **/build/test-results/**/TEST-*.xml
+            - |
+              echo "Job creation time: $((SEMAPHORE_JOB_CREATION_TIME * 1000))"
+              echo "Current time: $(date +%s%3N)"
+              echo $(( ($(date +%s%3N) - $SEMAPHORE_JOB_CREATION_TIME * 1000) )) > duration.txt
+              artifact push workflow duration.txt --destination $SEMAPHORE_JOB_INDEX-test-duration-$SEMAPHORE_WORKFLOW_ID || true
+
+  - name: Gradle Build
```
Comment (santhoshct) [inline .semaphore/semaphore.yml:81]: suggest making build as first block then followed by test.

Comment (Muen342) [reply .semaphore/semaphore.yml:56]: done

**Review by santhoshct — APPROVED** [review]
LGTM

**Pull request merged**